File: 백준/Gold/7569. 토마토/토마토.cc
#include <iostream>
#include <queue>
using namespace std;
struct tri{
  int a;
  int b;
  int c;
};
queue<tri> q;
int maxi=0;
int dx[6]={1,-1,0,0,0,0};
int dy[6]={0,0,1,-1,0,0};
int dh[6]={0,0,0,0,1,-1};
int arr[101][101][101];
int visited[101][101][101];
int m,n,h;
int main(){
  cin.tie(nullptr);
  cout.tie(nullptr);
  cin>>m>>n>>h;
  for(int i=0;i<h;i++){
    for(int j=0;j<n;j++){
      for(int k=0;k<m;k++){
        cin>>arr[i][j][k];
        if(arr[i][j][k]==1){
          q.push({i,j,k});
          visited[i][j][k]=1;
        }
      }
    }
  }
  // for(int i=0;i<h;i++){
  //   for(int j=0;j<n;j++){
  //     for(int k=0;k<m;k++){
  //       cout<<arr[i][j][k]<<" ";
  //     }
  //     cout<<endl;
  //   }
  // }
  while(!q.empty()){
    int first=q.front().a;
    int second=q.front().b;
    int third=q.front().c;

    q.pop();

    for(int l=0;l<6;l++){
      int nz=first+dh[l];
      int nx=second+dx[l];
      int ny=third+dy[l];

      if(nz>=0 && nx>=0 && ny>=0 && visited[nz][nx][ny]==0 && arr[nz][nx][ny]==0 && nz<h && nx<n && ny<m){
        q.push({nz, nx, ny});
        visited[nz][nx][ny]=1;
        arr[nz][nx][ny]=arr[first][second][third]+1;
        }
      }
  }
  for(int i=0;i<h;i++){
    for(int j=0;j<n;j++){
      for(int k=0;k<m;k++){
        if(arr[i][j][k]==0){
          cout<<-1;
          return 0;
        }
        else{
          maxi=max(maxi, arr[i][j][k]);
        }
      }
    }
  }
  cout<<maxi-1;
  // for(int i=0;i<h;i++){
  //   for(int j=0;j<n;j++){
  //     for(int k=0;k<m;k++){
  //       cout<<arr[i][j][k]<<" ";
  //     }
  //     cout<<endl;
  //   }
  // }
}

File: 백준/Gold/7569. 토마토/토마토.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main{
	public static int maxi = 0;
	public static int[] dx = {1,-1,0,0,0,0};
	public static int[] dy = {0,0,1,-1,0,0};
	public static int[] dz = {0,0,0,0,1,-1};
	public static int[][][] arr;
	public static boolean[][][] visited;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		List<Pair> l = new ArrayList<Pair>();
		String[] temp = br.readLine().split(" ");
		int m = Integer.parseInt(temp[0]);
		int n = Integer.parseInt(temp[1]);
		int h = Integer.parseInt(temp[2]);
		arr = new int[h][n][m];
		visited = new boolean[h][n][m];
		Queue<Pair> q = new LinkedList<Pair>();
		for(int i=0;i<h;i++) {
			for(int j=0;j<n;j++) {
				String[] docs = br.readLine().split(" ");
				for(int k=0;k<m;k++) {
					arr[i][j][k] = Integer.parseInt(docs[k]);
//					System.out.print(i+" "+j+" "+k+" ");
				}
			}
		}
		for(int i=0;i<h;i++) {
			for(int j=0;j<n;j++) {
				for(int k=0;k<m;k++) {
					if(arr[i][j][k]==1) {
						l.add(new Pair(i,j,k));
					}
				}
			}
		}
		
		for(Pair p: l) {
			q.add(new Pair(p.z, p.y, p.x));
			visited[p.z][p.y][p.x] = true;
			
		}
		
		while(!q.isEmpty()){
			Pair p = q.poll();
			int nop = p.z;
			int garo = p.x;
			int sero = p.y;
//			System.out.println(nop+" "+garo+" "+sero);
				for(int j=0;j<6;j++) {
					int nnop = nop+dz[j];
					int ngaro = garo+dx[j];
					int nsero = sero+dy[j];
					if(nnop>=0 && ngaro>=0 && nsero>=0 && nnop<h && ngaro<m && nsero<n && !visited[nnop][nsero][ngaro] && arr[nnop][nsero][ngaro] == 0) {
						arr[nnop][nsero][ngaro] = arr[nop][sero][garo] + 1;
						visited[nnop][nsero][ngaro] = true;
						q.add(new Pair(nnop, nsero, ngaro));
					}
				}
			
		}
		
		for(int i=0;i<h;i++) {
			for(int j=0;j<n;j++) {
				for(int k=0;k<m;k++) {
					if(arr[i][j][k]==0) {
						System.out.println(-1);
						System.exit(0);
					}
					else {
						maxi = Math.max(maxi, arr[i][j][k]);
					}
				}
			}
		}
		
		System.out.println(maxi-1);
		
	}
	public static class Pair{
		int z;
		int y;
		int x;
		Pair(int z, int y, int x) {
			this.z = z;
			this.y = y;
			this.x = x;
		}
	}
}
